src/lib/math/pcurves/pcurves_impl.h

```diff
@@ -0,0 +1,1404 @@
+            copy_mem(out, std::span{bytes}.template subspan<extra, Self::BYTES>());
+            return out;
+         }
+      }
+
+      template <size_t L>
+      std::array<W, L> stash_value() const {
+         static_assert(L >= N);
+         std::array<W, L> stash = {};
+         for(size_t i = 0; i != N; ++i) {
+            stash[i] = m_val[i];
+         }
+         return stash;
+      }
+
+      template <size_t L>
+      static Self from_stash(const std::array<W, L>& stash) {
+         static_assert(L >= N);
+         std::array<W, N> val = {};
+         for(size_t i = 0; i != N; ++i) {
+            val[i] = stash[i];
+         }
+         return Self(val);
+      }
+
+      // Returns nullopt if the input is an encoding greater than or equal P
+      constexpr static std::optional<Self> deserialize(std::span<const uint8_t> bytes) {
+         // We could allow either short inputs or longer zero padded
+         // inputs here, however it seems best to avoid non-canonical
+         // representations unless required
+         if(bytes.size() != Self::BYTES) {
+            return {};
+         }
+
+         const auto words = bytes_to_words<W, N, BYTES>(&bytes[0]);
+
+         if(!bigint_ct_is_lt(words.data(), N, P.data(), N).as_bool()) {
+            return {};
+         }
+
+         return Self::from_words(words);
+      }
+
+      // ECDSA style hash->scalar conversion
+      //
+      // This must accept inputs of any length
+      static Self from_bits_with_trunc(std::span<const uint8_t> bytes) {
+         const size_t bit_length = bytes.size() * 8;
+
+         if(bit_length <= Self::BITS) {
+            // No shifting required, but might still need to reduce by modulus
+            std::array<uint8_t, 2 * BYTES> padded_bytes = {};
+            copy_mem(padded_bytes.data() + 2 * BYTES - bytes.size(), bytes.data(), bytes.size());
+            return Self(Rep::wide_to_rep(bytes_to_words<W, 2 * N, 2 * BYTES>(&padded_bytes[0])));
+         } else {
+            const size_t shift = bit_length - Self::BITS;
+
+            if(shift % 8 == 0) {
+               // Easy case just copy different bytes
+               const size_t new_length = bytes.size() - (shift / 8);
+               return Self::from_bits_with_trunc(bytes.subspan(0, new_length));
+            } else {
+               // fixme
+               throw Not_Implemented("Bit shifting for hash to scalar conversion not implemented");
+            }
+         }
+      }
+
+      // Reduces large input modulo the order
+      template <size_t L>
+      static constexpr Self from_wide_bytes(std::span<const uint8_t, L> bytes) {
+         static_assert(8 * L <= 2 * Self::BITS);
+         std::array<uint8_t, 2 * BYTES> padded_bytes = {};
+         copy_mem(padded_bytes.data() + 2 * BYTES - L, bytes.data(), L);
+         return Self(Rep::wide_to_rep(bytes_to_words<W, 2 * N, 2 * BYTES>(&padded_bytes[0])));
+      }
+
+      static constexpr Self random(RandomNumberGenerator& rng) {
```

Comment:
Not sure about the `constexpr` here, because of `RNG::randomize()`.

```suggestion
      static Self random(RandomNumberGenerator& rng) {
```

Reply:
Oh indeed, thanks. I keep expecting the compiler will catch this for me.